pmlc/dialect/tile/transforms/pad_ranges.cc

```diff
@@ -0,0 +1,148 @@
+// Copyright 2019, Intel Corporation
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "llvm/Support/MathExtras.h"
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/padding.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/util/ident.h"
+#include "pmlc/util/logging.h"
+#include "pmlc/util/strides.h"
+
+namespace pmlc::dialect::tile {
+
+using namespace mlir; // NOLINT
+
+namespace {
+
+struct PadRangesPass : public PadRangesBase<PadRangesPass> {
+  int64_t maxPowerOfTwo = 16;
+  int64_t minPowerOfTwo = 8;
+  int64_t maxIncreasePercent = 5;
+  void runOnFunction() final;
+};
+
+void PadRangesPass::runOnFunction() {
+  auto func = getFunction();
+  func.walk([&](ContractionOp op) {
+    // Skip some cases where the padding pass can't operate.
+    if (op.getNumSymbols()) {
+      op.emitRemark("padding cannot be run on symbolic contractions");
+      return;
+    }
+    if (!op.lowerBounds() || !op.upperBounds()) {
+      op.emitRemark("contraction bounds must be computed");
+      return;
+    }
+    // Extract which dimensions of the contraction are which output dimension
```

Comment:
This function seems like a good candidate for a utility function as with `util::computeStrideArray` below.

Reply:
I think this is a somewhat unusual case as we only handle simple output dimensions (1 to 1), which is not normally the case.  I'd like to hold of extracting this to a utility until at leas one other use arises.